File: shared-lib/tests/c/events.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "tests.h"
#include "uhppoted.h"

bool getEventIndex() {
    uint32_t index;

    if (get_event_index(DEVICE_ID, &index) < 0) {
        printf("ERROR %s\n", errmsg());
        return false;
    }

    const result resultset[] = {
        {
            .field = "event index",
            .type = "uint32",
            .value.uint32.expected = 47,
            .value.uint32.value = index,
        },
    };

    return evaluate("get-event-index", sizeof(resultset) / sizeof(result), resultset);
}

bool setEventIndex() {
    const char *tag = "set-event-index";

    if (set_event_index(DEVICE_ID, EVENT_INDEX) < 0) {
        printf("ERROR %s\n", errmsg());
        return false;
    }

    const result resultset[] = {};

    return evaluate(tag, sizeof(resultset) / sizeof(result), resultset);
}

bool getEvent() {
    const char *tag = "set-event";
    event event;

    if (get_event(DEVICE_ID, EVENT_INDEX, &event) < 0) {
        printf("ERROR %s\n", errmsg());
        return false;
    }

    // clang-format off
    const result resultset[] = {
        { .field = "event timestamp", .type = "string",  .value.string.expected = "2022-04-15 12:29:15", .value.string.value = event.timestamp },
        { .field = "event index",     .type = "uint32",  .value.uint32.expected = 51,                    .value.uint32.value = event.index },
        { .field = "event type",      .type = "uint8",   .value.uint8.expected = 6,                      .value.uint8.value = event.eventType },
        { .field = "event granted",   .type = "boolean", .value.boolean.expected = true,                 .value.boolean.value = event.granted },
        { .field = "event door",      .type = "uint8",   .value.uint8.expected = 3,                      .value.uint8.value = event.door },
        { .field = "event direction", .type = "uint8",   .value.uint8.expected = 1,                      .value.uint8.value = event.direction },
        { .field = "event card",      .type = "uint32",  .value.uint32.expected = 8165538,               .value.uint32.value = event.card },
        { .field = "event reason",    .type = "uint8",   .value.uint8.expected = 21,                     .value.uint8.value = event.reason },
    };
    // clang-format on

    return evaluate(tag, sizeof(resultset) / sizeof(result), resultset);
}

bool recordSpecialEvents() {
    const char *tag = "record-special-events";

    if (record_special_events(DEVICE_ID, true) < 0) {
        printf("ERROR %s\n", errmsg());
        return false;
    }

    const result resultset[] = {};

    return evaluate(tag, sizeof(resultset) / sizeof(result), resultset);
}

File: shared-lib/go/events.go
//go:build !debug && !tests

package main

import (
	"C"
	"fmt"

	"github.com/uhppoted/uhppote-core/uhppote"
)

func getEventIndex(uu uhppote.IUHPPOTE, index *uint32, deviceID uint32) error {
	if index == nil {
		return fmt.Errorf("invalid argument (index) - expected valid pointer")
	}

	reply, err := uu.GetEventIndex(deviceID)
	if err != nil {
		return err
	}

	*index = reply.Index

	return nil
}

func setEventIndex(uu uhppote.IUHPPOTE, deviceID uint32, index uint32) error {
	reply, err := uu.SetEventIndex(deviceID, index)
	if err != nil {
		return err
	} else if !reply.Changed {
		return fmt.Errorf("%v: failed to set event index", deviceID)
	}

	return nil
}

func getEvent(uu uhppote.IUHPPOTE, event *C.struct_Event, deviceID uint32, index uint32) error {
	if event == nil {
		return fmt.Errorf("invalid argument (event) - expected valid pointer")
	}

	e, err := uu.GetEvent(deviceID, index)
	if err != nil {
		return err
	} else if e == nil {
		return fmt.Errorf("%v: no response to get-event %v", deviceID, index)
	}

	event.timestamp = C.CString(fmt.Sprintf("%v", e.Timestamp))
	event.index = C.uint(e.Index)
	event.eventType = C.uchar(e.Type)
	event.granted = cbool(e.Granted)
	event.door = C.uchar(e.Door)
	event.direction = C.uchar(e.Direction)
	event.card = C.uint(e.CardNumber)
	event.reason = C.uchar(e.Reason)

	return nil
}

func recordSpecialEvents(uu uhppote.IUHPPOTE, deviceID uint32, enabled bool) error {
	ok, err := uu.RecordSpecialEvents(deviceID, enabled)
	if err != nil {
		return err
	} else if !ok && enabled {
		return fmt.Errorf("%v: failed to enable record special events", deviceID)
	} else if !ok && !enabled {
		return fmt.Errorf("%v: failed to disable record special events", deviceID)
	}

	return nil
}
